Multinomial and Cox redux; also fixing showarray error

dsweber2 opened the pull request:
This is a minor rewrite of pull request https://github.com/JuliaStats/GLMNet.jl/pull/10 so that it actually works with the recent version of Julia.
I dropped the gadfly plots and dependency, as that would be better done using Recipes in the current Julia ecosystem.
I added `@testsets` to demarcate testign the different models.
In the process of all this I ran into a usage of the deprecated showarray, so I also fixed that.

JackDunnNZ reviewed (changes requested):
Thanks for reviving this @dsweber2! I had some very minor comments related to incorporating the few PRs that have happened here in the meantime

Given that it's not your work originally, I'm not sure how much effort you are inclined to put into this, but it's a bit of a shame that the original PR had a lot of duplication between the current code and the new files, the biggest ones that stuck out to me were:
- the duplication of the `GLMNetPath` struct, which could probably have a type parameter added instead so that it can reuse some of the generic functions
- the duplication of the `glmnetcv` function for cox
- duplication of the some of the generic functions like `predict`

I don't really have the time to go through and do this myself, but I don't want to block you if you need this added. It would be great if you were able to tidy it up, but I am happy to merge either way.

Another quick question I had, do we know if the results for the added tests line up with the results that R gives? It would be good to double check this if possible

dsweber2 commented on README.md:
```diff
@@ -17,63 +17,137 @@ julia> y = collect(1:100) + randn(100)*10;
-[53,]     2 0.911956 0.215546
-[54,]     2 0.911966 0.196397
-[55,]     2 0.911974  0.17895
+Least Squares GLMNet Solution Path (74 solutions for 4 predictors in 832 passes):
```
this would be easier if we had a Documenter.jl instead of a Readme file, though that's probably more overhead than either of us want/need to put in. Be that as it may, I went through and actually ran the code and made a commit with the new output. I also rewrote the plotting example so that it works with Plots.jl (no recipe, just a prototype plot in the documentation).

The file readmeExamples.jl has the actual code run that should be reproducible.

dsweber2 commented on src/GLMNet.jl:
```diff
@@ -532,16 +551,24 @@ function glmnetcv(X::AbstractMatrix, y::Union{AbstractVector,AbstractMatrix},
+    lambda = path.lambda[1:minimum(map(length, fits))]
+    fits = map(z->z[1:length(lambda)], fits)
+
+    fitloss = hcat(fits...)::Matrix{Float64}
```
oh whoops! I didn't notice they'd changed that as well, thanks for catching it. I've reverted it.

dsweber2 commented on test/runtests.jl:
```diff
@@ -102,6 +101,14 @@ path = glmnet(sparse(X), y)
+# plot(path, x = :norm2)
+# @test_throws ErrorException plot(path, x = :Coeff)
+# @test_throws ErrorException plot(path, y = :Coeff)
+
```
gone

dsweber2 commented on src/CoxNet.jl:
```diff
@@ -0,0 +1,241 @@
+            folds::Vector{Int}=begin
+                n, r = divrem(size(y, 1), nfolds)
+                shuffle!([repmat(1:nfolds, n); 1:r])
+            end, parallel::Bool=false, kw...)
```
also done, and added a check

dsweber2 commented:
I appreciate your quick turn-around time.
> the duplication of the GLMNetPath struct, which could probably have a type parameter added instead so that it can reuse some of the generic functions
this was easy enough to hack in through parametric types. There shouldn't be too many of them spawned by this, so it seemed a good use.
> the duplication of the glmnetcv function for cox
> duplication of the some of the generic functions like predict
I was under the impression that if they're sufficiently different in execution, using the type dispatch was the correct approach. I didn't really check if they could be integrated cleanly though, so I was mostly relying on the judgement of the original PR.

> Another quick question I had, do we know if the results for the added tests line up with the results that R gives? It would be good to double check this if possible

I haven't double checked it, and was assuming that the hand entered values were from R. I don't currently have glmnet working in R (part of the reason I hacked this together in the first place)

dsweber2 commented:
`CategoricalArray` was causing the 1.0 tests to break. I ditched it by getting more specific with the types in GLMNet.jl's copy of `glmnet` and making the `y` in Multinomial.jl generic. This didn't cause any tests to break, but it's not the most elegant solution.

I'm not sure what's going on with the broken tests now; check_jerr only has two arguments? the error claims it only has three...

JackDunnNZ commented:
Thanks for the changes, I think this is good to go once we get tests passing. I'm happy to trust that the magic numbers in the test make sense!

> I was under the impression that if they're sufficiently different in execution, using the type dispatch was the correct approach. I didn't really check if they could be integrated cleanly though, so I was mostly relying on the judgement of the original PR.

The only reason it stuck out to me was noticing that the cox version of glmnetcv needed to be updated with the `rng` change, and I guess in future they'll need to be kept in sync. No big deal, it's not like the package is in constant flux anyway.

> I'm not sure what's going on with the broken tests now; check_jerr only has two arguments? the error claims it only has three...

I'm guessing this branch needs to sync with master again, since I went through yesterday and merged a few old PRs and fixed a number of low-hanging issues. This error probably came from https://github.com/JuliaStats/GLMNet.jl/pull/51

dsweber2 commented:
For some reason, `CategoricalArray` wasn't showing up for the tests on the remote. It worked fine on my computer. So I added it as a test dependency, which seems to have solved the issue. Puzzling though, as I'm on 1.5.3 as well. I think we've addressed most of the extra stuff that came up. Thanks for keeping this maintained.

JackDunnNZ commented:
Thanks again, looks good to me!

Marco-Congedo commented:
Thanks guys to take the package up!
